File: Priority_Queue/1606. Find Servers That Handled Most Number of Requests/1606. Find Servers That Handled Most Number of Requests.cpp
class Solution {
public:
    vector<int> busiestServers(int k, vector<int>& arrival, vector<int>& load) {
        priority_queue<pair<int, int>, vector<pair<int,int>>, greater<pair<int,int>>>busy;
        set<int>free;
        vector<int>count(k, 0);
        
        for (int i = 0; i < k; i++) {
            free.insert(i);
        }
        
        for (int i = 0; i < arrival.size(); i++) {
            while (!busy.empty() && busy.top().first <= arrival[i]) {
                free.insert(busy.top().second);
                busy.pop();
            }
            
            if (free.empty()) continue;
            
            auto iter = free.lower_bound(i%k);
            if (iter == free.end()) {
                iter = free.begin();
            } 
            
            int serverId = *iter;
            free.erase(serverId);
            busy.push({arrival[i]+load[i], serverId});
            count[serverId]++;
        }
        int maxCount = *max_element(count.begin(), count.end());
        vector<int>rets;
        int idx = 0;
        for (int i = 0; i < k; i++) {
            if (count[i] != maxCount) continue;
            rets.push_back(i);
        }
        
        return rets;
    }
};

// free : sorted (set)
// busy : heap 

// emulate procees

// 1. set all elemtn int free;
// 2. tranverse arrival vector;
//     1) check whether busy has released id to free;
//     2) get conrrepindg id from free queue and delete it;
//     3) count the max val for every id;
// 3. according the max val to get the matched id

File: Priority_Queue/1606. Find Servers That Handled Most Number of Requests/1606. Find Servers That Handled Most Number of Requests.java
class Solution {
    public List<Integer> busiestServers(int k, int[] arrival, int[] load) {
        int N = arrival.length;
        PriorityQueue<int[]> busy = new PriorityQueue<>((a,b) -> (a[0]-b[0]));
        TreeSet<Integer> free = new TreeSet<>();
        int[] count = new int[k];
        int maxVal = 0;
        List<Integer> rets = new ArrayList<>();
        
        for (int i = 0; i < k; i++) {
            free.add(i);
        }
        
        for (int i = 0; i < N; i++) {
            while (!busy.isEmpty() && busy.peek()[0] <= arrival[i]) {
                int[] cur = busy.peek();
                busy.poll();
                free.add(cur[1]);
            }
            
            // find freed server
            if (free.isEmpty()) continue;
            int serverId;
            if (free.ceiling(i%k) != null) {
                serverId = free.ceiling(i%k);
            } else {
                // first server
                serverId= free.first();
            }
            free.remove(serverId);
            busy.offer(new int[] {arrival[i] + load[i], serverId});
            count[serverId]++;
            maxVal = Math.max(maxVal, count[serverId]);
        }
        
        for (int i = 0; i < k; i++) {
            if (count[i] == maxVal) rets.add(i);
        }
        return rets;
    }
}

// free -> sorted (TreeMap) 
// busy -> heap (time : id)
// emulate the processing

/*class Solution {
    public List<Integer> busiestServers(int k, int[] arrival, int[] load) {
        int[] servers = new int[k];
        int[] reqs = new int[k];
        int max = 0;
        
        TreeSet<Integer> available = new TreeSet<Integer>();
        for (int i=0; i<k; i++) {
            available.add(i);
        }
        PriorityQueue<int[]> busy = new PriorityQueue<int[]>((a, b) -> (a[0]-b[0]));
        
        for (int i=0; i<arrival.length; i++) {
            while (!busy.isEmpty() && busy.peek()[0] <= arrival[i]) {
                int[] server = busy.poll();
                available.add(server[1]);
            }
            if (available.isEmpty()) {
                continue;
            }
            Integer usable = available.ceiling(i%k);
            if (usable == null) {
                usable = available.first();
            }
            busy.offer(new int[]{arrival[i]+load[i], usable});
            available.remove(usable);
            reqs[usable] ++;
            if (max < reqs[usable]) {
                max = reqs[usable];
            }
        }
        
        List<Integer> busiest = new ArrayList<Integer>();
        for (int i=0; i<k; i++) {
            if (max == reqs[i]) busiest.add(i);
        }
        return busiest;
    }
}*/
